src/state/selectors.ts

```diff
@@ -26,29 +31,60 @@ function getSliced<T>(arr: ReadonlyArray<T>, startIndex: number, endIndex: numbe
   return arr;
 }
 
-export function selectTooltipPayload(state: RechartsRootState): TooltipPayload | undefined {
-  const { activeIndex, tooltipItemPayloads } = state.tooltip;
-  if (activeIndex === -1) {
-    return undefined;
-  }
-  const { chartData, dataStartIndex, dataEndIndex } = state.chartData;
-  /*
-   * If a payload has data specified directly from the graphical item, prefer that.
-   * Otherwise, fill in data from the chart level, using the same index.
-   */
-  return tooltipItemPayloads.map(({ dataDefinedOnItem, settings }) => {
-    const finalData = dataDefinedOnItem ?? chartData;
-
-    const sliced = getSliced(finalData, dataStartIndex, dataEndIndex);
-
-    // TODO settings coming from Tooltip props, and tooltipAxis
-    const tooltipPayload = sliced?.[activeIndex];
-    const tooltipEntry = getTooltipEntry({
-      tooltipEntrySettings: settings,
-      dataKey: settings?.dataKey,
-      payload: tooltipPayload,
-      value: getValueByDataKey(tooltipPayload, settings?.dataKey),
+const selectTooltipState = (state: RechartsRootState) => state.tooltip;
+const selectChartData = (state: RechartsRootState) => state.chartData;
+
+const selectTooltipTicks = createSelector(selectTooltipAxis, (tooltipAxis: BaseAxisProps) =>
+  getTicksOfAxis(tooltipAxis, false, true),
+);
+
+const selectActiveIndex = createSelector(selectTooltipState, (tooltipState: TooltipState) => tooltipState.activeIndex);
+
+const selectActiveLabel = createSelector(
+  selectTooltipTicks,
+  selectActiveIndex,
+  (tooltipTicks: ReadonlyArray<TickItem>, activeIndex: number): string | undefined =>
+    tooltipTicks?.[activeIndex]?.value,
+);
+
+export const selectTooltipPayload: (state: RechartsRootState) => TooltipPayload | undefined = createSelector(
+  selectTooltipState,
+  selectChartData,
+  selectTooltipAxis,
+  selectActiveLabel,
+  (
+    tooltipState: TooltipState,
+    chartDataState: ChartDataState,
+    tooltipAxis: BaseAxisProps | undefined,
+    activeLabel: string | undefined,
+  ) => {
+    const { activeIndex, tooltipItemPayloads } = tooltipState;
+    if (activeIndex === -1) {
+      return undefined;
+    }
+    const { chartData, dataStartIndex, dataEndIndex } = chartDataState;
+
+    /*
+     * If a payload has data specified directly from the graphical item, prefer that.
+     * Otherwise, fill in data from the chart level, using the same index.
+     */
+    return tooltipItemPayloads.map(({ dataDefinedOnItem, settings }) => {
+      const finalData = dataDefinedOnItem ?? chartData;
```

Comment:
this will need changed to a check for length or `dataDefinedOnItem` will need to be null/undefined

Reply:
Possible. Do we have all the cases covered with tests?

Reply:
no https://github.com/recharts/recharts/blob/3.x/src/cartesian/Scatter.tsx#L205
😅

Reply:
https://github.com/search?q=repo%3Arecharts%2Frecharts%20data%3A%20%5B%5D&type=code

Those first three there will all return `dataDefinedOnItem` every time

Reply:
That might actually explain some of the failures I saw.

Reply:
`Scatter` was returning nothing for Tooltip data. Now it is (once the dataKey issue is good to go too)

We also fix a bug where Scatter tooltips never get color styles when we do this

Reply:
And what's worse, the `getComposedData` will re-inject the `data` prop and overwrite whatever it was that was there before so we lose access to it.

Soooo a new blocker to this blocker, we need to remove `getComposedData` first.